schemas/dns.yml

```diff
@@ -0,0 +1,210 @@
+  fields:
+
+    - name: type
+      level: extended
+      type: keyword
+      short: The type of DNS event captured, query or answer.
+      description: >
+        The type of DNS event captured, query or answer.
+
+        If your source of DNS events only gives you DNS queries, you should only
+        create dns events of type `dns.type:query`.
+
+        If your source of DNS events gives you answers as well, you should create
+        one event per query (optionally as soon as the query is seen). And a
+        second event containing all query details as well as an array of answers.
+      example: answer
+
+    - name: id
+      level: extended
+      type: keyword
+      description: >
+        The DNS packet identifier assigned by the program that generated the
+        query. The identifier is copied to the response.
+      example: 62111
+
+    - name: op_code
+      level: extended
+      type: keyword
+      description: >
+        The DNS operation code that specifies the kind of query in the message.
+        This value is set by the originator of a query and copied into the
+        response.
+      example: QUERY
+
+    - name: flags
+      level: extended
+      type: keyword
+      short: Array of DNS flags.
+      description: >
+        Array of 2 letter DNS flags.
+
+        Expected values are: AA, TC, RD, RA, AD, CD, DO
+      example: [RD, RA]
+
+    - name: response_code
+      level: extended
+      type: keyword
+      description: The DNS response code.
+      example: NOERROR
+
+    - name: question.name
+      level: extended
+      type: keyword
+      short: The name being queried.
+      description: >
+        The name being queried.
+
+        If the name field contains non-printable
+        characters (below 32 or above 126), those characters should be represented
+        as escaped base 10 integers (\DDD). Back slashes and quotes should be escaped.
+        Tabs, carriage returns, and line feeds should be converted to \t, \r, and
+        \n respectively.
+      example: www.google.com
+
+    - name: question.type
+      level: extended
+      type: keyword
+      description: The type of record being queried.
+      example: AAAA
+
+    - name: question.class
+      level: extended
+      type: keyword
+      description: The class of of records being queried.
+      example: IN
+
+    - name: question.registered_domain
+      level: extended
+      type: keyword
+      short: The highest registered domain, stripped of the subdomain.
+      description: >
+        The highest registered domain, stripped of the subdomain.
+
+        For example, the registered domain for "foo.google.com" is "google.com".
+
+        This value can be determined precisely with a list like the public
+        suffix list (http://publicsuffix.org). Trying to approximate this by
+        simply taking the last two labels will not work well for TLDs such as "co.uk".
+      example: google.com
+
+    - name: question.length
```

Comment:
Do we need `question.length` as its own field? Getting the length of a field value is fairly easy to do with a scripted field, and with Painless it will be almost as fast as having it there in the first place.

E.g.:
```
  "script_fields": {
    "dns.question.length": {
      "script": {
        "source": "doc['dns.question.name'].value.length()"
      }
    }
  }
```

> This is useful to detect data exfiltration via DNS queries.

For detecting data exfiltration, the entropy (and maybe cardinality) of the subdomain should be more indicative than the length.

Reply:
Yes, I was adding it for the purpose of helping detect data exfiltration, and having a numeric field available for filtering or aggregating/bucketing with.

But your point about **subdomain** length being more useful than full question length for this is a good one.

I will simply take it out for now, in the interest of merging this PR sooner.

We do however have the field `dns.question.registered_domain`, so when computing it, a pipeline is already in a position to calculate the length of the subdomain. Perhaps as a follow-up PR we should add both `dns.question.subdomain` and `dns.question.subdomain_length`?

WDYT @cwurm @MikePaquette @andrewkroh ?

Reply:
Also of note in this discussion, Packetbeat has a field called `dns.question.size`. Here I was renaming it to `dns.question.length` (see the "pattern" discussion in the body of the PR).

So we're good with dropping it altogether?

Reply:
>> This is useful to detect data exfiltration via DNS queries.

> For detecting data exfiltration, the entropy (and maybe cardinality) of the subdomain should be more indicative than the length.

well, there are many things which could be considered as important (not just length), and I agree that it is easy to calculate somewhere. We do it in Logstash, I've provided config we use in first [PR](https://github.com/elastic/ecs/pull/180#issue-231825157). So if you believe that use case for `dns.question.length` is to narrow and shouldn't be in ECS — let it be. 

> But your point about subdomain length being more useful 

For sure, most of tools for DNS tunneling are using subdomain as a carrier for payload, but there are options to exfiltrate data within domain name itself (not subdomain), making queries to specific (controlled) DNS server. 

That's why I believe it make sense to left calculation of the entire query length instead of subdomain, if this case is still interesting for you.

Reply:
I'm personally ok with dropping `dns.question.length` from the standard since it can be easily calculated at search time. If users prefer to pre-calculate it, they are free to do so, but I don't think we should require it, and any standard applications (Beats, SIEM) using DNS data should calculate it rather than relying on it being present in the data.

I'm so-so about `dns.question.subdomain` and `dns.question.subdomain_length`. These can be calculated as well (subdomain is `dns.question.name` minus `dns.question.registered_domain`, subdomain_length can be calculated from that in turn).

We could think of including these fields and mark them as optional, even to provide the script that can be used to calculate them.

However, in the spirit of progress over perfection and incremental improvements, I'd vote for getting this PR in now with the minimum set of required fields so we can move on to other things, and later we can always come back to it if we see a burning need for more fields.

Reply:
@yugoslavskiy Makes sense about the domain name itself being used for data exfil. With the current minimal fieldset we have all the data in, and users/applications can choose how to look at it (whole question, just TLD, ETLD, subdomain(s), length of any of these). I think that should cover all common use cases, or do you see something that's missing?

Reply:
Yeah these fields not being required / present in the spec doesn't mean users can't add them as custom fields. Everyone's free to build on top of ECS.

Reply:
@cw Thanks. I won't get the question.subdomain PR started for now.